File: ChatGPT/ChatGPT/javascript/3-wayHandShake.js
const net = require('net');

// TCP 서버 생성
const server = net.createServer((socket) => {
    console.log('클라이언트가 연결되었습니다.');

    // 클라이언트와의 데이터 송수신
    socket.on('data', (data) => {
        console.log('클라이언트로부터 데이터 수신:', data.toString());

        // 3-way handshake 응답 전송
        if (data.toString() === 'SYN') {
            console.log('SYN 수신, ACK 전송');
            socket.write('ACK');
        } else if (data.toString() === 'ACK') {
            console.log('ACK 수신, 연결 완료');
        }
    });

    // 소켓 연결 종료 처리
    socket.on('close', () => {
        console.log('클라이언트와의 연결이 종료되었습니다.');
    });
});

// TCP 서버가 3000 포트에서 대기
const PORT = 3000;
server.listen(PORT, () => {
    console.log(`TCP 서버가 ${PORT} 포트에서 대기 중입니다.`);
});

File: ChatGPT/ChatGPT/go/3-wayHandShake.go
package main

import (
    "fmt"
    "net"
)

func main() {
    serverAddress := "127.0.0.1:12345"

    // TCP 소켓 열기
    listener, err := net.Listen("tcp", serverAddress)
    if err != nil {
        fmt.Println("Error listening:", err.Error())
        return
    }
    defer listener.Close()

    fmt.Println("Server listening on", serverAddress)

    // 클라이언트 연결 대기
    conn, err := listener.Accept()
    if err != nil {
        fmt.Println("Error accepting:", err.Error())
        return
    }
    defer conn.Close()

    fmt.Println("Connection established with", conn.RemoteAddr())

    // SYN 패킷 수신 및 처리
    synBuffer := make([]byte, 1024)
    _, err = conn.Read(synBuffer)
    if err != nil {
        fmt.Println("Error reading SYN:", err.Error())
        return
    }
    fmt.Println("SYN received:", string(synBuffer))

    // SYN-ACK 패킷 전송
    _, err = conn.Write([]byte("SYN-ACK"))
    if err != nil {
        fmt.Println("Error writing SYN-ACK:", err.Error())
        return
    }
    fmt.Println("SYN-ACK sent")

    // ACK 패킷 수신 및 처리
    ackBuffer := make([]byte, 1024)
    _, err = conn.Read(ackBuffer)
    if err != nil {
        fmt.Println("Error reading ACK:", err.Error())
        return
    }
    fmt.Println("ACK received:", string(ackBuffer))

    // 연결 완료 메시지 전송
    _, err = conn.Write([]byte("Connection established"))
    if err != nil {
        fmt.Println("Error writing connection message:", err.Error())
        return
    }
    fmt.Println("Connection established message sent")
}
